### ISSUE TITLE
Cadence/Temporal: Refactor proxies as shared libraries

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#1295]
@johncburns1 and I have decided that it's time to bite the bullet and do this refactor.  We've spent more time maintaining the old HTTP based communication mechanism than makes sense and that scheme is also preventing us from having workflow and activity operations be native C# tasks rather than the hacky future implementation.

I'm going to close related issues and link them to this and then write a preliminary spec below that we can discuss.

**Comment by jefflill** [comment]
Let's start out with modifying Cadence because we have other folks depending on it already and then follow up with Temporal.  They'll both work basically the same way.

Currently, our Cadence implementation consists of `CadenceClient` written in C# and a separate GOLANG (cadence-proxy) program that runs as a separate process that actually communicates with Cadence server via the native GO Cadence client.  The C# client and cadence-proxy each expose HTTP servers for sending and receiving messages to each other.  We went down this path 2 1/2 years ago because:

* The GOLANG Cadence client uses TChannel for communicating with Cadence server and this wasn't available on Windows (and we didn't want to go down that rabbit hole)

* There didn't seem to be any documentation describing the RPC protocol and state machines we'd need to implement.

Although our implementation is a bit convoluted, it has been used in production for a couple years for another project and has actually worked quite well.  It's time now to refresh this because:

1. We've been using ASP.NET Kestrel for hosting the C# side of the connection for .NET Core but this stopped working recently for later .NET SDKs (.NET Framework actually does continue to work because that is based on `WebListener`).  Fixing this is a high priority because we don't want o leave existing users high and dry.

2. Using HTTP based communication (as we implemented it) is somewhat problematic because there's no strict guarantees for the order in which requests will be processed when multiple requests are submitted in parallel due to how .NET `HttpClient` can reuse existing underlying socket connections.  This ended up being a challenge when the time came to ensure that workflow operations were always executed in the same order.

3. Expanding on item 2: We ended up having some C# client APIs return new future classes as part of the mechanism to ensure workflow operation ordering.  This ended up being unsatisfying because we had to invent the new future mechanism rather than using the standard C# `async Task` capabilities (actually, users end up having use our futures combined with standard async which is even more confusing).

4. Managing the proxies as a separate process and communicating over the network isn't ideal and it would be really nice if the proxy could be converted into a shared library that could be loaded into the user's app process where we'd communicate with function calls.

The rest of this comment discusses what I think we should do for each of Cadence and Temporal.

### Cadence

We've been planning on deprecating Cadence (but still doing minimal maintenance) and concentrating on Temporal going forward so the main thing we need to do is replacing the ASP.NET implementation with something that will work going forward.  The idea is to replace the HTTP protocol with lower level socket based communication (described further below).

So to make this as painless as possible, let's just replace the HTTP protocol with sockets but continue to have the proxy run as a separate process.  This means that nearly all of the combined code will remain and work exactly the same with only the low-level message send/receive code needing to change.

### Temporal

For Temporal, let's do the right thing and address all of the issues above by converting the proxy into a shared library and modifying the C# API by converting the future classes into standard async tasks.  Note that we'll still need to be able to run the proxy as a separate process so we can debug it separately in VS Code in pre-launched debug mode so we'll still need to implement a socket communication method in addition to the in-process function calls for the shared library implementation.

### Socket Protocol

**Do we really want to do this???**  The semantics will need to be exactly the same as the HTTP implementation.

**Comment by jefflill** [comment]
We're going to to drop the shared library thing as too hacky so I'm going to close this issue.  We've decided to leave the Cadence implementation as is and deprecate it and begin work on a native Temporal gRPC.

I'm also abandoning the effort to port these to use the EmbedIO webserver because I couldn't get it to work reliably.  The main problem is that it supports HTTP/1.0 instead of HTTP/1.1 out of the box which results in every proxy requesting opening a new socket connection resulting in socket buffer exhaustion.  I tried hacking the code to enable HTTP/1.1 and that worked better but it still wasn't reliable.

So I'm reverting these projects back to using ASP.NET Core for .NET Core environments and WebListener for .NET Framework and I'm targeting these at: net48, netcoreapp3.1 and net5.0, **but not netstandard2.0**.  We used to target netstandard2.0 which I believe was causing the conflicts.  The new **Neon.CSharp** code fixes the dynamic compilation problems resolving the **whack-a-mole** conundrum.
